### PR TITLE
cdp enhancement

### PR DESCRIPTION
**Pull request by wangjj9219**
1. support use lp token as collateral
2. provider call for pre-issuing stable coin to obtain more collateral and deposit into CDP, the collateral ratio will decrease.
3. provider call for selling collateral in CDP to obtain stable coin to repay debit, the collateral ratio will increase.

### CODE REVIEW

File: modules/cdp-engine/src/lib.rs [hunk 1]
```diff
@@ -827,6 +837,227 @@ impl<T: Config> Pallet<T> {
...
+		// debit cap check due to new issued stable coin
+		Self::check_debit_cap(currency_id, <LoansOf<T>>::total_positions(currency_id).debit)?;
+		Ok(())
+	}
+
+	/// Sell ​​the collateral locked in CDP to get stable coin to repay the debit,
+	/// and the collateral ratio will be increased. For single token collateral,
+	/// try to swap stable coin by DEX. For lp token collateral, try to remove liquidity
+	/// for lp token first, then swap the non-stable coin to get stable coin. If all
+	/// debti are repaid, the extra stable coin will be transferred directly back to
```
Comment (shaunxw) [inline modules/cdp-engine/src/lib.rs:951]:
```suggestion
	/// debit are repaid, the extra stable coin will be transferred directly back to
```

File: modules/cdp-engine/src/lib.rs [hunk 2]
```diff
@@ -827,6 +837,227 @@ impl<T: Config> Pallet<T> {
...
+
+				total_stable
+			}
+			_ => {
+				// swap collateral to stable coin
+				let limit = SwapLimit::ExactSupply(decrease_collateral, min_decrease_debit_value);
+				let swap_path = T::DEX::get_best_price_swap_path(
+					currency_id,
+					stable_currency_id,
+					SwapLimit::ExactSupply(decrease_collateral, min_decrease_debit_value),
```
Comment (shaunxw) [inline modules/cdp-engine/src/lib.rs:1026]:
```suggestion
					limit,
```

File: modules/cdp-engine/src/lib.rs [hunk 3]
```diff
@@ -911,68 +1142,117 @@ impl<T: Config> Pallet<T> {
...
+					} else {
+						(amount_1, token_0, amount_0)
+					};
+
+					// these's stable refund
+					if existing_stable > target_stable_amount {
+						<T as Config>::CDPTreasury::withdraw_collateral(
+							&who,
+							stable_currency_id,
+							existing_stable - target_stable_amount,
```
Comment (shaunxw) [inline modules/cdp-engine/src/lib.rs:1169]:
```suggestion
							existing_stable.checked_sub(target_stable_amount).expect("ensured existing stable amount greater than target; qed"),
```

**Review by shaunxw — APPROVED** [review]

**Comment by wangjj9219** [comment]
@qwer951123  @nnnnnnnnnni This pr contains changes about events related to collateral auction and loan position. Data services should be compatible with these changes.

**Pull request merged**